Add efficient evalpoly implementation for an AbstractVector of coefficients

Luapulu opened the pull request:
This implementation is about twice as fast as the current implementation for an `AbstractVector` of coefficients. See my benchmarks and a comparison with other implementations [here](https://discourse.julialang.org/t/i-have-a-much-faster-version-of-evalpoly-why-is-it-faster/79899?u=luapulu).

The implementation works with custom array indices. Because I use `@inbounds`, the AbstractVector subtype must, for safety reasons, guarantee that all integer indices greater than or equal to `firstindex` and smaller than or equal to `lastindex` are valid.

Luapulu committed:
Add efficient evalpoly implementation for an AbstractVector of coefficients

Luapulu committed:
Fix type instability

Luapulu commented:
Seems the errors have nothing to do with my changes here, but I'm not sure I'm qualified to rule it out.

stevengj commented on base/math.jl:
```diff
@@ -187,7 +187,22 @@ function evalpoly(x, p::Tuple)
+    # without the annotation v is not type stable in some cases,
+    # e.g. if p contains Float32 but x is Float64
+    T = promote_type(typeof(x), eltype(p))
+    @inbounds v::T = p[i]
```
Instead, you could do `v  = p[i] * one(eltype(p))`.  The compiler will optimize out the multiplication for hardware-floating-point types, and it will give you the type you want.

stevengj replied:
This `@inbounds` is not correct, because you haven't checked that `p` is nonempty.

Probably we should have a line:
```jl
isempty(p) && return zero(x) * one(eltype(p))
```
(The current `evalpoly` throws in this case.)

stevengj commented on base/math.jl:
```diff
@@ -187,7 +187,22 @@ function evalpoly(x, p::Tuple)
+        @inbounds v = muladd(v, x, p[i])
+        i -= 1
+    end
+
```
(You can probably omit all these blank lines in such a short function.)

stevengj commented on base/math.jl:
```diff
@@ -187,7 +187,22 @@ function evalpoly(x, p::Tuple)
 end
 
-evalpoly(x, p::AbstractVector) = _evalpoly(x, p)
+function evalpoly(x, p::AbstractVector)
```
This should probably replace `function _evalpoly(x, p)`, no?

Luapulu replied:
Yes, I just did some benchmarking for tuples and the while implementation seems to be as fast as the current `_evalpoly` version.

Luapulu committed:
Small alterations

- replace _evalpoly with new version
- remove blank lines
- deal with case where coefficient vector has length zero
- use nicer way to make variable v type stable
- add tests for zero length coefficients and evalpoly output type

Luapulu commented:
I agree with your comments and made some changes accordingly.

PatrickHaecker commented:
I quickly checked that [it's still faster](https://github.com/JuliaLang/julia/pull/56707#issuecomment-3728215620) on current master. Could you, @Luapulu, please sync your PR with master?
